File: api/src/main/java/org/apache/iceberg/DataFile.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.iceberg;

import static org.apache.iceberg.types.Types.NestedField.optional;
import static org.apache.iceberg.types.Types.NestedField.required;

import java.util.List;
import org.apache.iceberg.types.Types;
import org.apache.iceberg.types.Types.BinaryType;
import org.apache.iceberg.types.Types.IntegerType;
import org.apache.iceberg.types.Types.ListType;
import org.apache.iceberg.types.Types.LongType;
import org.apache.iceberg.types.Types.MapType;
import org.apache.iceberg.types.Types.StringType;
import org.apache.iceberg.types.Types.StructType;

/** Interface for data files listed in a table manifest. */
public interface DataFile extends ContentFile<DataFile> {
  // fields for adding delete data files
  Types.NestedField CONTENT =
      optional(
          134,
          "content",
          IntegerType.get(),
          "Contents of the file: 0=data, 1=position deletes, 2=equality deletes");
  Types.NestedField FILE_PATH =
      required(100, "file_path", StringType.get(), "Location URI with FS scheme");
  Types.NestedField FILE_FORMAT =
      required(101, "file_format", StringType.get(), "File format name: avro, orc, or parquet");
  Types.NestedField RECORD_COUNT =
      required(103, "record_count", LongType.get(), "Number of records in the file");
  Types.NestedField FILE_SIZE =
      required(104, "file_size_in_bytes", LongType.get(), "Total file size in bytes");
  Types.NestedField COLUMN_SIZES =
      optional(
          108,
          "column_sizes",
          MapType.ofRequired(117, 118, IntegerType.get(), LongType.get()),
          "Map of column id to total size on disk");
  Types.NestedField VALUE_COUNTS =
      optional(
          109,
          "value_counts",
          MapType.ofRequired(119, 120, IntegerType.get(), LongType.get()),
          "Map of column id to total count, including null and NaN");
  Types.NestedField NULL_VALUE_COUNTS =
      optional(
          110,
          "null_value_counts",
          MapType.ofRequired(121, 122, IntegerType.get(), LongType.get()),
          "Map of column id to null value count");
  Types.NestedField NAN_VALUE_COUNTS =
      optional(
          137,
          "nan_value_counts",
          MapType.ofRequired(138, 139, IntegerType.get(), LongType.get()),
          "Map of column id to number of NaN values in the column");
  Types.NestedField LOWER_BOUNDS =
      optional(
          125,
          "lower_bounds",
          MapType.ofRequired(126, 127, IntegerType.get(), BinaryType.get()),
          "Map of column id to lower bound");
  Types.NestedField UPPER_BOUNDS =
      optional(
          128,
          "upper_bounds",
          MapType.ofRequired(129, 130, IntegerType.get(), BinaryType.get()),
          "Map of column id to upper bound");
  Types.NestedField KEY_METADATA =
      optional(131, "key_metadata", BinaryType.get(), "Encryption key metadata blob");
  Types.NestedField SPLIT_OFFSETS =
      optional(
          132, "split_offsets", ListType.ofRequired(133, LongType.get()), "Splittable offsets");
  Types.NestedField EQUALITY_IDS =
      optional(
          135,
          "equality_ids",
          ListType.ofRequired(136, IntegerType.get()),
          "Equality comparison field IDs");
  Types.NestedField SORT_ORDER_ID =
      optional(140, "sort_order_id", IntegerType.get(), "Sort order ID");
  Types.NestedField SPEC_ID = optional(141, "spec_id", IntegerType.get(), "Partition spec ID");


  StructType columnStatsType = StructType.of(
          required(152, "column_id", LongType.get(), "Column ID"),
          required(153, "subcolumn_path", StringType.get(), "UTF-8 String of subcolumn path"),
          required(154, "subcolumn_data_type", StringType.get(), "data type of this subcolumn for this file"),
          required(155, "column_value_count", LongType.get(), "column value count"),
          required(156, "column_null_value_count", LongType.get(), "column null value count"),
          required(157, "column_nan_value_count", LongType.get(), "column nan value count"),
          required(158, "column_distinct_count", LongType.get(), "column distinct value count"),
          required(159, "column_lower_bound", LongType.get(), "column lower bound value"),
          required(160, "column_upper_bound", LongType.get(), "column upper bound value"));
  Types.NestedField COLUMN_STATS =
          optional(
                  162, "column_stats", ListType.ofRequired(161, columnStatsType), "Column stats");


  int PARTITION_ID = 102;
  String PARTITION_NAME = "partition";
  String PARTITION_DOC = "Partition data tuple, schema based on the partition spec";

  // NEXT ID TO ASSIGN: 142

  static StructType getType(StructType partitionType) {
    // IDs start at 100 to leave room for changes to ManifestEntry
    return StructType.of(
        CONTENT,
        FILE_PATH,
        FILE_FORMAT,
        SPEC_ID,
        required(PARTITION_ID, PARTITION_NAME, partitionType, PARTITION_DOC),
        RECORD_COUNT,
        FILE_SIZE,
        COLUMN_SIZES,
        VALUE_COUNTS,
        NULL_VALUE_COUNTS,
        NAN_VALUE_COUNTS,
        LOWER_BOUNDS,
        UPPER_BOUNDS,
        KEY_METADATA,
        SPLIT_OFFSETS,
        EQUALITY_IDS,
        SORT_ORDER_ID);
  }

  /**
   * @return the content stored in the file; one of DATA, POSITION_DELETES, or EQUALITY_DELETES
   */
  @Override
  default FileContent content() {
    return FileContent.DATA;
  }

  @Override
  default List<Integer> equalityFieldIds() {
    return null;
  }
}

File: python_legacy/iceberg/api/data_file.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from iceberg.api.types import (BinaryType,
                               IntegerType,
                               ListType,
                               LongType,
                               MapType,
                               NestedField,
                               StringType,
                               StructType)


class DataFile(object):

    @staticmethod
    def get_type(partition_type):
        return StructType.of([NestedField.required(100, "file_path", StringType.get()),
                             NestedField.required(101, "file_format", StringType.get()),
                             NestedField.required(102, "partition", partition_type),
                             NestedField.required(103, "record_count", LongType.get()),
                             NestedField.required(104, "file_size_in_bytes", LongType.get()),
                             NestedField.required(105, "block_size_in_bytes", LongType.get()),
                             NestedField.optional(106, "file_ordinal", IntegerType.get()),
                             NestedField.optional(107, "sort_columns", ListType.of_required(112, IntegerType.get())),
                             NestedField.optional(108, "column_sizes", MapType.of_required(117, 118,
                                                                                           IntegerType.get(),
                                                                                           LongType.get())),
                             NestedField.optional(109, "value_counts", MapType.of_required(119, 120,
                                                                                           IntegerType.get(),
                                                                                           LongType.get())),
                             NestedField.optional(110, "null_value_counts", MapType.of_required(121, 122,
                                                                                                IntegerType.get(),
                                                                                                LongType.get())),
                             NestedField.optional(125, "lower_bounds", MapType.of_required(126, 127,
                                                                                           IntegerType.get(),
                                                                                           BinaryType.get())),
                             NestedField.optional(128, "upper_bounds", MapType.of_required(129, 130,
                                                                                           IntegerType.get(),
                                                                                           BinaryType.get()))]
                             # NEXT ID TO ASSIGN: 131
                             )

    def path(self):
        raise NotImplementedError()

    def format(self):
        raise NotImplementedError()

    def partition(self):
        raise NotImplementedError()

    def record_count(self):
        raise NotImplementedError()

    def file_size_in_bytes(self):
        raise NotImplementedError()

    def block_size_in_bytes(self):
        raise NotImplementedError()

    def file_ordinal(self):
        raise NotImplementedError()

    def sort_columns(self):
        raise NotImplementedError()

    def column_sizes(self):
        raise NotImplementedError()

    def value_counts(self):
        raise NotImplementedError()

    def null_value_counts(self):
        raise NotImplementedError()

    def lower_bounds(self):
        raise NotImplementedError()

    def upper_bounds(self):
        raise NotImplementedError()

    def copy(self):
        raise NotImplementedError()
